pulp_ansible/app/tasks/collections.py

```diff
@@ -205,3 +134,218 @@ def _update_highest_version(collection_version):
+            metadata = await get_json_from_url(url=get_page_url(remote.url))
+
+            page_count = math.ceil(float(metadata["count"]) / float(PAGE_SIZE))
+            progress_bar.total = page_count
+            progress_bar.save()
+
+            yield metadata
+            progress_bar.increment()
+
+            # Concurrent downloads are limited by aiohttp...
+            not_done = set(
+                asyncio.ensure_future(get_json_from_url(url=get_page_url(remote.url, page)))
+                for page in range(2, page_count + 1)
+            )
+
+            while not_done:
+                done, not_done = await asyncio.wait(not_done, return_when=asyncio.FIRST_COMPLETED)
+                for item in done:
+                    yield item.result()
+                    progress_bar.increment()
+
+
+class AnsibleContentSaver(ContentSaver):
+    """
+    A modification of ContentSaver stage that additionally saves Ansible plugin specific items.
+
+    Saves UpdateCollection, UpdateCollectionPackage, UpdateReference objects related to
+    the UpdateRecord content unit.
+    """
+
+    async def _pre_save(self, batch):
+        """
+        Save a batch of Collection objects.
+
+        Args:
+            batch (list of :class:`~pulpcore.plugin.stages.DeclarativeContent`): The batch of
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects to be saved.
+
+        """
+        for d_content in batch:
+            if d_content is None:
+                continue
+            if not isinstance(d_content.content, CollectionVersion):
+                continue
+
+            info = d_content.content.natural_key_dict()
+            collection, created = Collection.objects.get_or_create(
+                namespace=info["namespace"], name=info["name"]
+            )
+            d_content.content.collection = collection
+
+    async def _post_save(self, batch):
+        """
+        Save a batch of CollectionVersion, Tag objects.
+
+        Args:
+            batch (list of :class:`~pulpcore.plugin.stages.DeclarativeContent`): The batch of
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects to be saved.
+
+        """
+        for d_content in batch:
+            if d_content is None:
+                continue
+            if not isinstance(d_content.content, CollectionVersion):
+                continue
+            collection_version = d_content.content
+            for d_artifact in d_content.d_artifacts:
+                artifact = d_artifact.artifact
+                with tarfile.open(str(artifact.file.path), "r") as tar:
+                    log.info(_("Reading MANIFEST.json from {path}").format(path=artifact.file.path))
+                    file_obj = tar.extractfile("MANIFEST.json")
+                    manifest_data = json.load(file_obj)
+                    info = manifest_data["collection_info"]
+
+                    # Create the tags
```

Comment:
There is an additional requirement that just was introduced today from another PR (sorry). Please have the sync code preserve the call to [_update_hightest_value()](https://github.com/pulp/pulp_ansible/blob/master/pulp_ansible/app/tasks/collections.py#L190). See how if you rebase on the sync machinery you would be preserving the call removed here:  https://github.com/pulp/pulp_ansible/blob/master/pulp_ansible/app/tasks/collections.py#L178

Without that the `is_highest` field would be out of date when sync brings in newer highest versions.

Reply:
it is already there, just before `collection_version.save()`

Reply:
Awesome! Thank you @fabricio-aguiar, I agree.